simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/StationState.java

```diff
@@ -5,33 +5,36 @@
 import com.evbox.everon.ocpp.simulator.station.evse.Connector;
 import com.evbox.everon.ocpp.simulator.station.evse.Evse;
 import com.google.common.collect.ImmutableList;
-import lombok.AllArgsConstructor;
 
 import java.time.*;
 import java.util.List;
 import java.util.Optional;
 
 import static com.evbox.everon.ocpp.v20.message.station.StatusNotificationRequest.ConnectorStatus.AVAILABLE;
 import static java.util.stream.Collectors.toList;
-import static org.apache.commons.lang3.StringUtils.isBlank;
 import static org.apache.commons.lang3.StringUtils.isNotBlank;
 
-@AllArgsConstructor
+/**
+ * Contains all station data that can be mutated.
+ * Allowed to be accessed only by 'station-consumer' thread.
+ * The rest of threads can use getView() method for informational and testing purposes.
+ * @see StationMessageConsumer
+ */
```

Comment:
I think mentioning thread-names and method names is not mature. Cant we simply make this class thread-safe? That will give us confidence.

Reply:
It is not thread-safe by design, in order to provide simplicity of working with state within handlers and other classes that are part of the station's thread. Right now there's no way for other threads to get access to state (station doesn't have such API), this comment was added as a precaution for those who wouldn't understand framework and would like to give direct access to state in the future.

Reply:
What prevents us to make it thread-safe? It will be much easier to understand and you dont have to maintain javadocs.

Reply:
Please provide the solution, I don't understand what you want. It's a state that's supposed to mutate. To protect it from concurrent mutations we have queue. That's how the whole framework is done here. Otherwise we'll need to redesign bunch of things

Reply:
Why not just simply use a RW lock in station state (and in evse)? You don't have to refresh at all in that case (or worry when to refresh), you just block on read.

Reply:
What I'm saying is we can make the solution mature. We could use java.util.concurrent.locks.Lock  mechanism to provide thread-safety. For example you could use ReentrantReadWriteLock. Evse could also use Lock if its needed. 
In this case we dont need to do refreshing. No need to refactor other parts of the code.
The bug description states: Station state has to be thread-safe
Right now we can get list of evses directly (getEvses() or getEvseIds()) and its not thread-safe.

Reply:
We'll have to use lock for main object, each evse, each connector and other objects in the future that can be mutated inside of state. I don't see how it's simpler, instead of not worry about concurrent usage at all.

Reply:
I dont say that its simpler. I say that the solution should be mature. Currently the state is not thread-safe. And the bug is about that.

Reply:
State is not accessed by multiple threads anymore. There's no need for everything to be thread-safe

Reply:
Do we want to have readable and maintainable code? I believe yes, no need to refresh where its not needed, both of the reviewers suggest to revisit this solution because it can be improved.
IF there is no need then dont make it thread-safe, but in this case the solution is not mature and we suggest how to make it better. That also impacts extendability of this class, if you have a mature mechanism in place any developer can extend it faster. Right now its not clear unless if you read some javadocs what is called from which threads. And when I created this Bug I meant this. To make it thread-safe so that we dont think about it in the future.

Reply:
The problem is it's not some standard data-structure that you make thread-safe once and forget about it. You'll have to think about thread-safety each time new features are being added to it. Even if we make state thread-safe - if we expose it to other threads - it will be broken. By design state can be changed only by 1 thread at the time. Look at AvailabilityManager#determineEvseStatus for instance: Thread A mutates state based on the current state. If at any time thread B mutates state - next mutations which are done by Thread A will be invalid. It means that using lock per operation is not enough, it has to be transactional (multiple operations per lock acquisition).

Reply:
Thats the point of introducing a common pattern how to make it thread-safe, for example RRWL. You just follow this pattern if you add new methods. And dont forget about Reentrancy when you have multiple operations :-)

Reply:
Each caller will have to acquire lock in this case, everywhere, before accessing the state in any way. I disagree that it will be easier than current implementation

Reply:
Well, if current implementation raises concerns between us (developers who are familiar with code-base) then for developers who are not familiar it will be even harder.
Do we need to refresh every time when we receive a new request? Even if the request does not change anything in the state?

Reply:
Code base is not trivial, I agree. If I'd add some lock-based framework - it wouldn't be less confusing. 

Yes, we'll need to refresh copy each time, having some flag as 'dirtyCopy' seems unnecessary complication to me, refreshing copy on each write seems too complex as well and raises question whether we want to have copy of intermediate state. In my head I draw an analogy with 2-phase commit. We do many actions in scope of single message handling and create new snapshot after all actions have been performed.

Reply:
Would you agree if I implement this code and then we can see if its confusing or not?

I think the code should do the logic if its needed. Otherwise its a bad design.

Reply:
Of course I agree